### DISCUSSION TITLE
v4.0.0-beta

### DISCUSSION THREAD

**Discussion by rechsteiner** [rechsteiner/Parchment#667] [General]
**Version 4.0** introduced a new and improved API for PageView in SwiftUI. Each page can now show define their own menu items, which can be any SwiftUI view. The new API also supports result builders, which allows you to easily setup your PageView. See more in the [README](https://github.com/rechsteiner/Parchment/tree/4.0).

```swift
PageView {
    Page("Title 1") {
        Text("Page 1")
    }
    Page { _ in
        Image(systemName: "star.fill")
    } content: {
        Text("Page 2")
    }
}
```

## Changes
- New API for PageView in SwiftUI #666
- Updated minimum deployment target to iOS 11 #661

## Added
- Add horizontal right alignment support #658

<hr /><em>This discussion was created from the release <a href='https://github.com/rechsteiner/Parchment/releases/tag/v4.0.0-beta'>v4.0.0-beta</a>.</em>

**Comment by AlexandreOliveiraCarvalho** [comment]
Hello @rechsteiner, could you provide an approximate timeline for the deployment of version 4.0.0? Thank you.

**Reply by rechsteiner** [reply]
Hi @AlexandreOliveiraCarvalho! I haven't found any issues with the beta, so it's pretty much ready to go. I just need to set aside some time to document the new API a bit better, and I also want to check a few things related to the new Observation changes in SwiftUI.

**Reply by iletai** [reply]
Same as me, I follow the Beta version and hope it's as soon as RC. Should we need an open release for the hearing community about issues with the iOS version, performance, or something else?

**Reply by rechsteiner** [reply]
Sorry that this is taking some time. There's some issues in the current version because of the way we handle SwiftUI updates. With the new API, it's not easy to actually track if something has changed or not, and we end up reloading the underlying page view even when things have not changed. This causes issues with things like navigation where the scroll view gets reset, not to mention that it's not great for performance. I'm trying to find a proper solution, but we might need to change the API a little to include some kind of identifiers.

**Reply by iletai** [reply]
> Sorry that this is taking some time. There's some issues in the current version because of the way we handle SwiftUI updates. With the new API, it's not easy to actually track if something has changed or not, and we end up reloading the underlying page view even when things have not changed. This causes issues with things like navigation where the scroll view gets reset, not to mention that it's not great for performance. I'm trying to find a proper solution, but we might need to change the API a little to include some kind of identifiers.

Should we have existing issues or some demo for this? I want to follow also. Hope everyone can help something @rechsteiner

**Reply by rechsteiner** [reply]
I've updated the 4.0 branch now, which should address the issues that's been blocking this release https://github.com/rechsteiner/Parchment/commit/279c9355fbf478cb7e06b4ef8ed514d346f75f7c. The commit message contains more details on the potential fix. I'm still not sure this is the best approach, so I would like to continue testing this for a bit longer. Please let me know if you find any issues with the latest branch, or if you have any other ideas of solving this 🙏

**Comment by Raegonoth** [comment]
Hey @rechsteiner, I found a bug in the beta (I didn't check prod, so maybe there too).
When you wrap the PageView inside a NavigationStack/NavigationView and present it in a sheet, most of the time it removes the NavigationBar and only shows the PageView. Maybe the bug is with Apple, but it never happened to me when using sheets with anything else. My Target is iOS 16, here is the testCode:
`import Parchment
import SwiftUI
import UIKit

struct DefaultView: View {
    var body: some View {
		Text("Test")
			.sheet(isPresented: .constant(true)) {
				NavigationStack {
					PageView {
						Page { _ in
							Image(systemName: "star.fill")
								.padding()
						} content: {
							Text("Page 1")
								.font(.largeTitle)
								.foregroundColor(.gray)
						}

						Page("Title 2") {
							Text("Page 2")
								.font(.largeTitle)
								.foregroundColor(.gray)
						}

						Page("Title 3") {
							Text("Page 3")
								.font(.largeTitle)
								.foregroundColor(.gray)
						}

						Page("Title 4") {
							Text("Page 4")
								.font(.largeTitle)
								.foregroundColor(.gray)
						}

						Page("Some very long title") {
							Text("Page 5")
								.font(.largeTitle)
								.foregroundColor(.gray)
						}

						Page("Title 6") {
							Text("Page 6")
								.font(.largeTitle)
								.foregroundColor(.gray)
						}

						Page("Title 7") {
							Text("Page 7")
								.font(.largeTitle)
								.foregroundColor(.gray)
						}
					}.navigationTitle("Test")
				}
			}
    }
}`

**Reply by rechsteiner** [reply]
Thanks for the report! This bug seems to be related to the way we handle SwiftUI updates. I don't have a proper solution in place yet, but hoping to have something ready soon.

**Reply by Raegonoth** [reply]
Hey @rechsteiner did you fix it with 4.0.0? Because it looks like it! 💪🏻

**Reply by rechsteiner** [reply]
Yep 🙌 It seems to be working fine now, and haven't heard about any issues

**Comment by ashutosh-allen** [comment]
I was using the tag: `v4.0.0-beta`, in the class `PageViewCoordinator`, there are compilation error coming up, for `Cannot find UIViewController in the scope`, I think this is because UIKit is not imported in that file. Also can you help, when this tag will be published?

**Reply by rechsteiner** [reply]
Hi @ashutosh-allen! I'll look into getting this fixed. I'm don't have any date for when this release will be ready, as we still need to solve some issues with the new approach (see https://github.com/rechsteiner/Parchment/discussions/667#discussioncomment-6174913).

**Reply by rechsteiner** [reply]
@ashutosh-allen can you try targeting the latest 4.0 branch and see if it works now?
